plugin/call_hierarchy.py

```diff
@@ -0,0 +1,157 @@
+from .core.promise import Promise
+from .core.protocol import CallHierarchyIncomingCall
+from .core.protocol import CallHierarchyIncomingCallsParams
+from .core.protocol import CallHierarchyItem
+from .core.protocol import CallHierarchyOutgoingCall
+from .core.protocol import CallHierarchyOutgoingCallsParams
+from .core.protocol import CallHierarchyPrepareParams
+from .core.protocol import Request
+from .core.registry import new_tree_view_sheet
+from .core.registry import windows
+from .core.registry import get_position
+from .core.registry import LspTextCommand
+from .core.registry import LspWindowCommand
+from .core.tree_view import TreeDataProvider
+from .core.tree_view import TreeItem
+from .core.typing import cast
+from .core.typing import IntEnum, List, Optional
+from .core.views import parse_uri
+from .core.views import SYMBOL_KINDS
+from .core.views import text_document_position_params
+from functools import partial
+import sublime
+
+
+class CallHierarchyDirection(IntEnum):
+    IncomingCalls = 1
+    OutgoingCalls = 2
+
+
+class CallHierarchyDataProvider(TreeDataProvider):
+
+    def __init__(
+        self,
+        window: sublime.Window,
+        session_name: str,
+        direction: CallHierarchyDirection,
+        root_elements: List[CallHierarchyItem]
+    ) -> None:
+        self.window = window
+        self.session_name = session_name
```

Comment:
It might not be the only place that has this issue but I feel that it might not be correct to use `session_name` for matching as there might be multiple sessions with same config name in a single window. Maybe let's instead create a weak reference to session (`weakref.ref(session)`) and use that?

Reply:
I use a weakref for the CallHierarchyDataProvider now https://github.com/sublimelsp/LSP/pull/2151/commits/fbfc6cc67c6ca6afa17d2090b4866e03b427a6b0. But I think the same is not possible for the "toggle direction" command, because you can only pass basic JSON-compatible values in the [command args](https://www.sublimetext.com/docs/api_reference.html#sublime.CommandArgs).

Reply:
Might be worth investigating adding a unique ID to session and lookup through that. But not in this PR of course.